src/Compilers/CSharp/Portable/Symbols/DelegateCaches/TypeOrMethodScopedDelegateCacheContainer.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    using FieldSymbolsCollection = SmallDictionary<ValueTuple<TypeSymbol, MethodSymbol>, FieldSymbol>;
+
+    internal class TypeOrMethodScopedDelegateCacheContainer : DelegateCacheContainer
+    {
+        private readonly Symbol _containingSymbol;
+        public override Symbol ContainingSymbol => _containingSymbol;
+
+        private int _delegateFields;
+        private readonly FieldSymbolsCollection _delegateFieldsDict = new FieldSymbolsCollection();
+
+        public override Accessibility DeclaredAccessibility => Accessibility.Private;
+
+        private readonly string _name;
+        public override string Name => _name;
+
+        /// <summary>
+        /// Contains the mapping from current method's type parameters to this container's type parameters.
+        /// Only used by method scoped generic container.
+        /// </summary>
+        private readonly TypeMap _typeMap;
+
+        private readonly ImmutableArray<TypeParameterSymbol> _typeParameters;
+        public override ImmutableArray<TypeParameterSymbol> TypeParameters => _typeParameters;
+
+        /// <summary>Creates a type scoped concrete delegate cache container.</summary>
+        internal TypeOrMethodScopedDelegateCacheContainer(TypeSymbol containingType, int generation)
+            : base(DelegateCacheContainerKind.TypeScopedConcrete)
+        {
+            Debug.Assert(containingType.IsDefinition);
+
+            _name = GeneratedNames.MakeDelegateCacheContainerName(0, generation, null);
+            _containingSymbol = containingType;
+            _typeParameters = ImmutableArray<TypeParameterSymbol>.Empty;
+        }
+
+        /// <summary>Creates a method scoped generic delegate cache container.</summary>
+        internal TypeOrMethodScopedDelegateCacheContainer(MethodSymbol currentMethod, int methodOrdinal, int generation)
+            : base(DelegateCacheContainerKind.MethodScopedGeneric)
+        {
+            Debug.Assert(currentMethod.IsDefinition);
+
+            _name = GeneratedNames.MakeDelegateCacheContainerName(methodOrdinal, generation, currentMethod.Name);
+            _containingSymbol = currentMethod.ContainingType;
+            _typeMap = TypeMap.Empty.WithAlphaRename(currentMethod, this, out _typeParameters);
+        }
+
+        internal override FieldSymbol ObtainCacheField(SyntheticBoundNodeFactory factory, TypeSymbol delegateType, MethodSymbol targetMethod)
```

Comment:
delegateType is a NamedTypeSymbol, parameter should reflect that as well as type arguments of FieldSymbolsCollection.